File: problem-117/solution.go
package problem117

import "container/list"

type treeNode struct {
	val         int
	left, right *treeNode
}

type queue struct {
	l *list.List
}

type queueItem struct {
	node  *treeNode
	level int
}

func newQueue() *queue {
	l := list.New()
	q := &queue{
		l: l,
	}
	return q
}

func (q *queue) push(node *treeNode, level int) {
	item := queueItem{node: node, level: level}
	q.l.PushBack(item)
}

func (q *queue) pop() (node *treeNode, level int) {
	e := q.l.Front()
	q.l.Remove(e)
	item := e.Value.(queueItem)
	return item.node, item.level
}

func (q *queue) isEmpty() bool {
	return q.l.Len() == 0
}

func minLevelSum(root *treeNode) int {
	if root == nil {
		return 0
	}

	nodeQueue := newQueue()
	nodeQueue.push(root, 0)
	sumByLevel := make(map[int]int)

	for !nodeQueue.isEmpty() {
		currNode, currLevel := nodeQueue.pop()
		sumByLevel[currLevel] += currNode.val

		if currNode.left != nil {
			nodeQueue.push(currNode.left, currLevel+1)
		}

		if currNode.right != nil {
			nodeQueue.push(currNode.right, currLevel+1)
		}
	}

	minLevel, minSum := 0, sumByLevel[0]
	for level, sum := range sumByLevel {
		if sum < minSum {
			minSum = sum
			minLevel = level
		}
	}

	return minLevel
}

File: problem-117/solution.py
from typing import Deque, Dict, Tuple
from collections import defaultdict, deque


class TreeNode:
    def __init__(self, val: int, left: 'TreeNode' = None, right: 'TreeNode' = None) -> None:
        self.val = val
        self.left = left
        self.right = right


def minimum_level_sum(root: TreeNode) -> int:
    if not root:
        return 0

    queue: Deque[Tuple[TreeNode, int]] = deque()
    queue.append((root, 0))
    sum_by_level: Dict[int, int] = defaultdict(int)

    while queue:
        node, curr_level = queue.popleft()
        sum_by_level[curr_level] += node.val
        if node.left:
            queue.append((node.left, curr_level + 1))
        if node.right:
            queue.append((node.right, curr_level + 1))

    min_level, min_sum = 0, sum_by_level[0]
    for level, sum in sum_by_level.items():
        if sum < min_sum:
            min_level = level

    return min_level

test_node = TreeNode(15, TreeNode(1, TreeNode(4)), TreeNode(2, TreeNode(6), TreeNode(8)))
assert minimum_level_sum(test_node) == 1